python_modules/libraries/dagster-k8s/dagster_k8s/container_context.py

```diff
@@ -204,6 +215,201 @@ def get_environment_dict(self) -> Mapping[str, str]:
+                )
+
+        validated_container_context = self
+
+        if only_allow_user_defined_env_vars is not None:
+            validated_container_context = self._filter_user_defined_env_vars(
+                set(only_allow_user_defined_env_vars)
+            )
+
+        return validated_container_context
+
+    def _filter_user_defined_k8s_config(
+        self,
+        user_defined_k8s_config: UserDefinedDagsterK8sConfig,
+        only_allow_user_defined_env_vars: Set[str],
+        discarded_keys: Set[str],
+    ) -> UserDefinedDagsterK8sConfig:
+        if "env" not in user_defined_k8s_config.container_config:
+            return user_defined_k8s_config
+
+        filtered_env = []
+        for env_dict in user_defined_k8s_config.container_config["env"]:
+            env_key = env_dict["name"]
+            if env_key in only_allow_user_defined_env_vars:
+                filtered_env.append(env_dict)
+            else:
+                discarded_keys.add(env_key)
+
+        return UserDefinedDagsterK8sConfig.from_dict(
+            {
+                **user_defined_k8s_config.to_dict(),
+                "container_config": {
+                    **user_defined_k8s_config.container_config,
+                    "env": filtered_env,
+                },
+            }
+        )
+
+    def _filter_user_defined_env_vars(
+        self,
+        only_allow_user_defined_env_vars: Set[str],
+    ) -> "K8sContainerContext":
+        discarded_keys = set()
+
+        new_run_k8s_config = self._filter_user_defined_k8s_config(
+            self.run_k8s_config,
+            only_allow_user_defined_env_vars,
+            discarded_keys,
+        )
+
+        new_server_k8s_config = self._filter_user_defined_k8s_config(
+            self.server_k8s_config,
+            only_allow_user_defined_env_vars,
+            discarded_keys,
+        )
+
+        new_env_vars = []
+        for env_var in self.env_vars:
+            env_var_name = get_env_var_name(env_var)
+            if env_var_name in only_allow_user_defined_env_vars:
+                new_env_vars.append(env_var)
+            else:
+                discarded_keys.add(env_var_name)
+
+        new_env = []
+        for env_map in self.env:
+            if env_map["name"] in only_allow_user_defined_env_vars:
+                new_env.append(env_map)
+            else:
+                discarded_keys.add(env_map["name"])
+
+        if discarded_keys:
+            logging.warning(
+                f"Excluding the following environment variables because they are not in the allowlist for user-defined environment variables: {', '.join(discarded_keys)}"
+            )
+
+        return self._replace(
+            env_vars=new_env_vars,
+            env=new_env,
+            run_k8s_config=new_run_k8s_config,
+            server_k8s_config=new_server_k8s_config,
+        )
+
+    def _get_used_k8s_config_fields(self) -> Mapping[str, Mapping[str, Set[str]]]:
+        used_fields = {}
```

Comment:
is it worth worrying about the exhaustive-ness of this impl? Like new or missed fields not getting caught here would then end up not triggering the error right?

Reply:
its a fair question - the various cases here mostly cover top-level fields. I think now that we have server_k8s_config and raw_k8s_config as an escape hatch, new top level fields should be quite rare.